### ISSUE TITLE
Regex performance problem in V8

### ISSUE THREAD

**Issue by mathiasbynens** [twitter/twemoji#84]
See https://code.google.com/p/chromium/issues/detail?id=482998.

Note that the optimization described in https://code.google.com/p/chromium/issues/detail?id=482998#c6 that avoids this problem comes for free if you use [Regenerate](https://mths.be/regenerate) to generate that part of the regular expression.

**Comment by WebReflection** [comment]
Hey Mathias ... are you opening a bug about V8 performance? :D

I don't think V8 related bugs should compromise or change this library. I think V8 engineers should figure out the issue (also, about ranges ... for an immutable RegExp? I'd expect V8 to figure that out but regardless I've tried ranges in twemoji RegExp already before first publication and it wasn't smaller, neither faster, across all engines)

So, what should I do exactly here?

**Comment by mathiasbynens** [comment]
It’s up to you of course, but if you wanted you could optimize the regex to avoid this V8-specific problem. The resulting output would also be smaller (in byte size) than the current regexp.

If you don’t care then feel free to close this issue. :)

**Comment by WebReflection** [comment]
I'm not sure I should solve a v8 problem not because I don't care about performance but because others might have the same issue. Should we all write RegExp the v8 or maybe they can fix this issue?

If it's "unfixable" then of course I'll try to bring ranges in the RegExp but I'd like to avoid the introduction of external libraries, if possible.

**Comment by mathiasbynens** [comment]
> If it's "unfixable" then of course I'll try to bring ranges in the RegExp but I'd like to avoid the introduction of external libraries, if possible.

Note that Regenerate is not a run-time dependency (although you could use it like that, but I don’t recommend it), only a build tool (i.e. `devDependencies`).

**Comment by WebReflection** [comment]
Does this RegExp look good to you ?

```js
/((?:[\xA9\xAE\u2106\u2113-\u211C\u2122\u23E9-\u23EC\u23F0\u23F3\u26CE\u2705\u270A\u270B\u2728\u274C\u274E\u2753-\u2755\u2795-\u2797\u27B0\u27BF\u3030\uE50A]|\uD82D[\uDC3D-\uDC7B\uDC7D\uDC7F-\uDD34\uDD36-\uDD39\uDD3D-\uDD7A\uDD8D-\uDDA4\uDDAC\uDDAD\uDDBA\uDDCA\uDDCD-\uDDD6\uDE22-\uDE7D\uDE82-\uDE8D\uDEBD-\uDF02\uDF3C-\uDF5C\uDF6C-\uDF71\uDF73-\uDFB8\uDFBC-\uDFCF\uDFDC-\uDFFF]|\uD82E[\uDC00\uDC02-\uDC06\uDC1C-\uDC2C]|\uD89B[\uDC4B\uDC53\uDC5A\uDC5C\uDC5F\uDC69\uDC6A])|(?:(?:[\u203C\u2049\u2139\u2194-\u2199\u21A9\u21AA\u231A\u231B\u24C2\u25AA\u25AB\u25B6\u25C0\u25FB-\u25FE\u2600\u2601\u260E\u2611\u2614\u2615\u261D\u263A\u2648-\u2653\u2660\u2663\u2665\u2666\u2668\u267B\u267F\u2693\u26A0\u26A1\u26AA\u26AB\u26BD\u26BE\u26C4\u26C5\u26D4\u26EA\u26F2\u26F3\u26F5\u26FA\u26FD\u2702\u2708\u2709\u270C\u270F\u2712\u2714\u2716\u2733\u2734\u2744\u2747\u2757\u2764\u27A1\u2934\u2935\u2B05-\u2B07\u2B1B\u2B1C\u2B50\u2B55\u303D\u3297\u3299]|\uD82D[\uDC40\uDDBB\uDE56\uDE6B])([\uFE0E\uFE0F]?)))/g
```

**Comment by mathiasbynens** [comment]
To be honest, it would be easier to review the script that generated it :)

**Comment by WebReflection** [comment]
OK, I answer by my own. NO, it doesn't work.

Somehow I've got '\ufe0f\u20e3' and '\u20e3' non matched, so tests are not green anymore if I use the `regenerate` version of the RegExp.

These are OSX specific, non standard, variants. Any idea if these could be fixed/solved somehow ?

**Comment by WebReflection** [comment]
here the branch ... not sure I am missing something, done different tests/attempt to have non standard OSX emoji in too but `test.html` keeps failing

https://github.com/twitter/twemoji/blob/regenerate/twemoji-generator.js#L216-L244

**Comment by rodrigopolo** [comment]
Hi, I’m not a JavaScript expert, but after listing all emojis available on iOS 8.3 and reviewing how they are encoded in JavaScript, also, after reviewing the Unicode.org emoji-data.txt I have found that the simple way to have a regex that can handle all emojis is by having all emojis in a regex using an OR condition, maybe I’m wrong but in practice it works.

The trick I did to prevent catching emojis that are made by several emojis is done by sorting the emoji list by length before creating the regex, you can see this code/workaround here:

https://github.com/rodrigopolo/minEmoji/blob/master/jMinEmoji2/js/jMinEmoji2.js

You can see the complete emoji list here:

http://rodrigopolo.com/files/emojilist/

I am a pragmatic person, that’s why instead of following all the standard and guidelines I focus on iOS 8.3 and OS X 10.10.3 because that’s where most people type emojis, the list is based on all emoji images existent in “Apple Color Emoji.ttf”, if the image isn’t there is quite impossible that someone typed that emoji anywhere on the web or any app, I also included in the list all Unicode and UTF8 JavaScript codes, the description available in emoji-data.txt and the description (or short code) found in emoji-cheat-sheet.com.

You can download the JavaScript where all emojis are listed here:
http://rodrigopolo.com/files/emojilist/table.js

Notice that I have added some new diverse emoji to the list and I have added some SVG by changing the color of the original from twemoji, 300+

I hope this helps in some way to your work.

**Comment by Otto42** [comment]
Note that the issue I posted at https://code.google.com/p/chromium/issues/detail?id=482998 is now marked there as fixed, and I can confirm the fix in Chrome 45, currently in the beta channel. So, this might not actually be an issue in V8 anymore.
